### PR TITLE
Emit ATen ops when symbolics raise + minor fixes

### LINKED ISSUE #75623: [ONNX] `ONNX_ATEN_FALLBACK` supports ONNX Runtime
Previously setting `operator_export_type=OperatorExportTypes.ONNX_ATEN_FALLBACK` produed a model that was only usable by Caffe2.
This issue tracks work to make it usable by ONNX Runtime as well.

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Currently `torch.onnx.export(.., operator_export_type=OperatorExportTypes.ONNX_ATEN_FALLBACK)` only issues ATen ops through explicit requests (e.g. `g.at()`) calls inside each op symbolic function. This is done based on specific conditions such as `operator_export_type==OperatorExportTypes.ONNX_ATEN_FALLBACK)` or `is_caffe2_aten_fallback()`

This PR extends the ATen fallback mechanism for scenarios when the symbolic function raises `RuntimeError` during export. The idea is that partial implementation of existing ONNX ops can fallback to ATen as a last resort. That is valuable because each operator can have many input combinations and not all are always implemented. 

A minor fix was done to make sure the `overload_name` attribute is added to explicit ATen op fallback requests when a symbolic is not registered to a particular op.

ps: The behavior for builds with BUILD_CAFFE2=1 is not changed to ensure BC.

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
Must be merged after https://github.com/pytorch/pytorch/pull/73954

File: test/onnx/test_operators.py [hunk 1]
```diff
@@ -352,6 +355,46 @@ def test_clip(self):
...
+                            opset_version=_onnx_opset_version)
+        finally:
+            unregister_custom_op_symbolic('::clamp', _onnx_opset_version)
+
+    @skipIfCaffe2
+    def test_clip_aten_fallback_explicit_request(self):
+        from torch.onnx import symbolic_registry
+
+        def is_registered_op_tampered(opname, domain, version):
+            fake_missing_symbolics = ('clamp')
```
Comment (garymm) [inline test/onnx/test_operators.py:380]: This is not creating a tuple, I'm pretty sure this is a bug.
You need a trailing comma `("clamp",)`

File: test/onnx/test_operators.py [hunk 2]
```diff
@@ -352,6 +355,46 @@ def test_clip(self):
...
+            if opname in fake_missing_symbolics:
+                return False
+            return (domain, version) in symbolic_registry._registry \
+                and opname in symbolic_registry._registry[(domain, version)]
+
+        try:
+            # Monkey patch to force missing symbolic for well-known op
```
Comment (garymm) [inline test/onnx/test_operators.py:387]: can you use unittest.mock instead of doing this manual try / finally stuff?

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -1161,12 +1170,21 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
             attrs = {k + "_" + n.kindOf(k)[0]: n[k] for k in n.attributeNames()}
             outputs = n.outputsSize()
             attrs["outputs"] = outputs
-            return g.at(op_name, *inputs, **attrs)
+            # `overload_name` is set for non-Caffe2 builds only
+            return g.at(op_name, *inputs, overload_name=_get_aten_op_overload_name(n), **attrs)
         else:
             raise sym_registry.UnsupportedOperatorError(domain, op_name, opset_version)
     except RuntimeError:
+        # TODO: Should RuntimeError be used for unsupported operators?
```
Comment (garymm) [inline torch/onnx/utils.py:1178]: I think a custom exception type makes sense but I would file an issue in GitHub rather than adding a TODO comment here.

**Review by garymm — DISMISSED** [review]
Thanks!

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -385,6 +397,19 @@ def run_model_test_with_external_data(self, model, input, rtol=0.001, atol=1e-7,
                 ort_outs = run_ort(ort_sess, input_copy)
                 ort_compare_with_pytorch(ort_outs, output, rtol, atol)
 
+    def assert_is_aten_op(self, onnx_model, operator, overload_name=""):
+        all_aten_nodes = [p for p in onnx_model.graph.node if p.name == "ATen" and p.domain == "org.pytorch.aten"]
+        if not all_aten_nodes:
+            return False
```
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:403]:
```suggestion
        assert all_aten_nodes
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -385,6 +397,18 @@ def run_model_test_with_external_data(self, model, input, rtol=0.001, atol=1e-7,
...
+        assert all_aten_nodes
+
+        for op in all_aten_nodes:
+            attrs = {attr.name: attr for attr in op.attribute}
+            if attrs.get("operator") == operator:
+                break
+        assert op.name == operator
+        assert attrs.get("operator") == operator
+        assert attrs.get("overload_name", "") == overload_name
+        return True
```
Comment (thiagocrepaldi) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:411]:
```suggestion
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -82,7 +86,7 @@ def convert_to_onnx(model, input=None, opset_version=9, do_constant_folding=True
                     keep_initializers_as_inputs=True, dynamic_axes=None,
                     input_names=None, output_names=None,
                     fixed_batch_size=False, training=None,
-                    verbose=False):
+                    verbose=False, operator_export_type=torch.onnx.OperatorExportTypes.ONNX):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:89]: Add `from torch.onnx import OperatorExportTypes` at the top to make other references to it much shorter.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 7]
```diff
@@ -10762,6 +10785,45 @@ def forward(self, x):
         self.run_test(Module(False), x, rtol=1e-3, atol=1e-6)
         self.run_test(Module(True), x, rtol=1e-3, atol=1e-6)
 
+    @skipIfCaffe2
+    def test_clip_aten_fallback_due_exception(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:10789]: can you move these to `test/jit/test_onnx_export.py`? These tests don't use onnxruntime, so it's confusing to have them in this file.

**Review by garymm — CHANGES_REQUESTED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 8]
```diff
@@ -82,7 +82,7 @@ def convert_to_onnx(model, input=None, opset_version=9, do_constant_folding=True
                     keep_initializers_as_inputs=True, dynamic_axes=None,
                     input_names=None, output_names=None,
                     fixed_batch_size=False, training=None,
-                    verbose=False):
+                    verbose=False, operator_export_type=torch.onnx.OperatorExportTypes.ONNX):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:85]: can we revert all changes in this file now?

File: torch/testing/_internal/common_utils.py [hunk 9]
```diff
@@ -2424,6 +2425,19 @@ def assertGreaterAlmostEqual(self, first, second, places=None, msg=None, delta=N
         msg = self._formatMessage(msg, standardMsg)
         raise self.failureException(msg)
 
+    def assert_is_aten_op(self, onnx_model, operator, overload_name=""):
+        all_aten_nodes = [p for p in onnx_model.graph.node
+                          if p.op_type == "ATen" and p.domain == "org.pytorch.aten"]
+        self.assertTrue(len(all_aten_nodes) > 0)
```
Comment (garymm) [inline torch/testing/_internal/common_utils.py:2431]:
```suggestion
        self.assertTrue(all_aten_nodes)
```

File: torch/testing/_internal/common_utils.py [hunk 10]
```diff
@@ -2424,6 +2425,19 @@ def assertGreaterAlmostEqual(self, first, second, places=None, msg=None, delta=N
...
+
+        for op in all_aten_nodes:
+            attrs = {attr.name: attr.s.decode() for attr in op.attribute}
+            if attrs.get("operator") == operator:
+                break
+
+        self.assertEqual(attrs.get("operator"), operator)
```
Comment (garymm) [inline torch/testing/_internal/common_utils.py:2438]: can operator be None?
If not, then replace with
```suggestion
        self.assertEqual(attrs["operator"], operator)
```

**Review by garymm — APPROVED** [review]
Thanks!
Please address comments before merging.

**Comment by thiagocrepaldi** [comment]
@garymm @bowangbj It seems that the pipelines that run `test/jit/test_onnx_export.py` does not support `onnx` (either not installed or maybe an ancient version)? 

```
======================================================================
ERROR [0.080s]: test_clip_aten_fallback_due_exception (jit.test_onnx_export.TestONNXExport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1072, in wrapper
    return func(self)
  File "/var/lib/jenkins/workspace/test/jit/test_onnx_export.py", line 395, in test_clip_aten_fallback_due_exception
    onnx_model = onnx.load_from_string(f.getvalue())
AttributeError: module 'onnx' has no attribute 'load_from_string'

======================================================================
ERROR [0.018s]: test_clip_aten_fallback_explicit_request (jit.test_onnx_export.TestONNXExport)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1072, in wrapper
    return func(self)
  File "/var/lib/jenkins/workspace/test/jit/test_onnx_export.py", line 417, in test_clip_aten_fallback_explicit_request
    onnx_model = onnx.load_from_string(f.getvalue())
AttributeError: module 'onnx' has no attribute 'load_from_string'

----------------------------------------------------------------------
Ran 2644 tests in 122.392s

FAILED (errors=2, skipped=130, expected failures=13)
```

Have you seen this before? It does work on my machine and the test also worked on `test/onnx/test_pytorch_onnx_onnxruntime.py`

File: test/onnx/test_onnx_export.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
+# Owner(s): ["module: onnx"]
```
Comment (garymm) [inline test/onnx/test_onnx_export.py:1]: I'm OK with this for now but I'll want to combine it with this file once it's merged:
https://github.com/pytorch/pytorch/pull/73284/files#diff-630243becc710c122152dc52e75109e86fc92a505fb293584e47ad3e4031fb65

File: test/onnx/test_operators.py [hunk 12]
```diff
@@ -5,7 +5,9 @@
 
 import torch
 import torch.onnx
-from torch.onnx.symbolic_helper import parse_args, _get_tensor_dim_size, _get_tensor_sizes
```
Comment (garymm) [inline test/onnx/test_operators.py:8]: please revert changes to this file.

File: scripts/onnx/test.sh [hunk 13]
```diff
@@ -75,7 +75,9 @@ if [[ "$BUILD_ENVIRONMENT" == *ort_test1* ||  "${SHARD_NUMBER}" == "1" ]]; then
     "$top_dir/test/onnx/test_utility_funs.py" \
     "$top_dir/test/onnx/test_pytorch_onnx_caffe2.py" \
     "$top_dir/test/onnx/test_pytorch_onnx_caffe2_quantized.py" \
-    "$top_dir/test/onnx/test_pytorch_onnx_shape_inference.py"
+    "$top_dir/test/onnx/test_pytorch_onnx_shape_inference.py" \
+    "$top_dir/test/onnx/test_onnx_export.py"
+
```
Comment (garymm) [inline scripts/onnx/test.sh:80]: remove empty line

**Review by garymm — APPROVED** [review]
Thanks!

**Review by msaroufim — APPROVED** [review]

**Pull request closed without merging**